<a href="https://colab.research.google.com/github/MuthiahAinun/Proyek_Analisis_Sentimen/blob/main/Skema2_Proyek_Analisis_Sentimen_Tsamarah_Muthi'ah_A.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Proyek Analisis Sentimen: [Youtube-Comments-dataset]
- **Nama:** [Tsamarah Muthi'ah Abdullah]
- **Email:** [a135xaf486@devacademy.id]
- **ID Dicoding:** [a135xaf48]

# Import Library

In [ ]:
# Menginstall google API
!pip install --upgrade google-api-python-client pandas

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 89.9/89.9 kB 2.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.1/13.1 MB 75.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.1/13.1 MB 78.4 MB/s eta 0:00:00
  Attempting uninstall: pandas
    Found existing installation: pandas 2.2.2
    Uninstalling pandas-2.2.2:
      Successfully uninstalled pandas-2.2.2
  Attempting uninstall: google-api-python-client
    Found existing installation: google-api-python-client 2.160.0
    Uninstalling google-api-python-client-2.160.0:
      Successfully uninstalled google-api-python-client-2.160.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
google-colab 1.0.0 requires pandas==2.2.2, but you have pandas 2.2.3 which is incompatible.


In [3]:
!pip install gensim

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.0/61.0 kB 3.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.6/60.6 kB 3.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 26.7/26.7 MB 60.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 18.3/18.3 MB 76.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 38.6/38.6 MB 15.2 MB/s eta 0:00:00
  Attempting uninstall: numpy
    Found existing installation: numpy 2.0.2
    Uninstalling numpy-2.0.2:
      Successfully uninstalled numpy-2.0.2
  Attempting uninstall: scipy
    Found existing installation: scipy 1.14.1
    Uninstalling scipy-1.14.1:
      Successfully uninstalled scipy-1.14.1


In [5]:
!pip install --upgrade numpy

In [6]:
!pip install --upgrade gensim

  Using cached numpy-1.26.4-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (61 kB)
Using cached numpy-1.26.4-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (18.3 MB)
  Attempting uninstall: numpy
    Found existing installation: numpy 2.2.4
    Uninstalling numpy-2.2.4:
      Successfully uninstalled numpy-2.2.4


In [1]:
# Import Library
import googleapiclient.discovery
import pandas as pd
import time
import random
import re
import string
import tensorflow as tf
import numpy as np
from tensorflow.keras.layers import Embedding, LSTM, Dense, Dropout, Conv1D, GlobalMaxPooling1D, Bidirectional, BatchNormalization
from sklearn.metrics import classification_report, accuracy_score
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, LSTM, GRU, Conv1D, GlobalMaxPooling1D, Dense, Dropout, Bidirectional, BatchNormalization
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from imblearn.over_sampling import SMOTE
from gensim.models import Word2Vec

# Data Preprocessing dan Pembersihan

In [2]:
# Membaca dan Menampilkan informasi dataset
dataset = pd.read_csv("NDsO1LT_0lw_youtube_comments.csv")
print("Jumlah data:", len(dataset))

Jumlah data: 85028


In [3]:
print(dataset.head())

             username                                               text  \
0            @MrBeast  BEAST GAMES FINALE DROPS FEBRUARY 13TH! GO WAT...   
1     @SultanRayyan-7                          Saya orang Indonesia 🇮🇩🇮🇩   
2  @MohamadrezaRezayy  مستر بیست عزیز ❤ لطفا دوبله فارسی این ویدیو را...   
3       @ILoveyou-954  বাংলাদেশ আসবেন কবে পাবনা জেলার কিছু দেখতে চাই ...   
4     @Moneymakerarab  This is for girls listen to Dalida - Helwa ya ...   

   likes          published_at  
0  54600  2025-02-08T16:59:31Z  
1      0  2025-03-19T21:51:42Z  
2      0  2025-03-19T21:43:03Z  
3      0  2025-03-19T21:34:32Z  
4      0  2025-03-19T21:20:21Z  


In [4]:
print("Informasi Dataset:")
print(dataset.info())

Informasi Dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85028 entries, 0 to 85027
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   username      84982 non-null  object
 1   text          85028 non-null  object
 2   likes         85028 non-null  int64 
 3   published_at  85028 non-null  object
dtypes: int64(1), object(3)
memory usage: 2.6+ MB
None


In [5]:
print('\nStatistik Deskriptif:')
print(dataset.describe())


Statistik Deskriptif:
               likes
count   85028.000000
mean        9.727654
std       744.702565
min         0.000000
25%         0.000000
50%         0.000000
75%         1.000000
max    167348.000000


In [6]:
# Mengecek jumlah missing values di setiap kolom
print(dataset.isnull().sum())

username        46
text             0
likes            0
published_at     0
dtype: int64


In [7]:
# Menghapus nilai kosong
dataset.dropna(inplace=True)

In [8]:
# Menghapus duplikasi
dataset.drop_duplicates(subset=['text'], inplace=True)

In [10]:
# Pembersihan teks
def clean_text(text):
    text = re.sub(r'@\w+', '', text)  # Menghapus username
    text = re.sub(r'http\S+', '', text)  # Menghapus URL
    text = re.sub(r'[^A-Za-z\s]', '', text)  # Menghapus karakter non-huruf
    text = text.lower().strip()  # Mengubah ke huruf kecil dan menghapus spasi
    return text

dataset['cleaned_text'] = dataset['text'].apply(clean_text)
print("Data setelah pembersihan:")
print(dataset.head())

Data setelah pembersihan:
             username                                               text  \
0            @MrBeast  BEAST GAMES FINALE DROPS FEBRUARY 13TH! GO WAT...   
1     @SultanRayyan-7                          Saya orang Indonesia 🇮🇩🇮🇩   
2  @MohamadrezaRezayy  مستر بیست عزیز ❤ لطفا دوبله فارسی این ویدیو را...   
3       @ILoveyou-954  বাংলাদেশ আসবেন কবে পাবনা জেলার কিছু দেখতে চাই ...   
4     @Moneymakerarab  This is for girls listen to Dalida - Helwa ya ...   

   likes          published_at  \
0  54600  2025-02-08T16:59:31Z   
1      0  2025-03-19T21:51:42Z   
2      0  2025-03-19T21:43:03Z   
3      0  2025-03-19T21:34:32Z   
4      0  2025-03-19T21:20:21Z   

                                        cleaned_text  
0  beast games finale drops february th go watch ...  
1                               saya orang indonesia  
2                                                     
3                                                     
4  this is for girls listen to dalida

In [11]:
# Mengecek kembali jumlah missing values di setiap kolom
print(dataset.isnull().sum())

username        0
text            0
likes           0
published_at    0
cleaned_text    0
dtype: int64


# **2. Ekstraksi Fitur dan Pelabelan Data**

Metode yang digunakan bebas sesuai dengan preferensi masing-masing peserta. Tahapan ini penting untuk mempersiapkan data sehingga dapat diolah lebih lanjut dalam proses pelatihan model.

In [12]:
# Melabeli komentar secara manual (positif, netral, negatif)
def label_comment(text):
    if any(word in text for word in ["love", "great", "awesome", "good", "nice", "amazing", "fantastic"]):
        return "positif"
    elif any(word in text for word in ["hate", "bad", "awful", "worst", "terrible", "disgusting", "boring"]):
        return "negatif"
    else:
        return "netral"

dataset['label'] = dataset['cleaned_text'].apply(label_comment)
print("Distribusi label:")
print(dataset['label'].value_counts())

Distribusi label:
label
netral     63770
positif     4724
negatif       69
Name: count, dtype: int64


In [14]:
# Encoding label
label_encoder = LabelEncoder()
dataset['label_encoded'] = label_encoder.fit_transform(dataset['label'])

# One-Hot Encoding pada label
y = tf.keras.utils.to_categorical(dataset['label_encoded'], num_classes=len(label_encoder.classes_))

print("Label Setelah Encoding:")
print(y[:5])
print("Classes:", label_encoder.classes_)

Label Setelah Encoding:
[[0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]
Classes: ['negatif' 'netral' 'positif']


In [15]:
# Tokenizer dan Padding
tokenizer = Tokenizer(num_words=10000, oov_token="<OOV>")
tokenizer.fit_on_texts(dataset['cleaned_text'])
sequences = tokenizer.texts_to_sequences(dataset['cleaned_text'])
padded = pad_sequences(sequences, maxlen=100, padding='post')

In [16]:
# Mengatasi Ketidakseimbangan Data dengan SMOTE
smote = SMOTE(random_state=42)
X_resampled, y_resampled = smote.fit_resample(padded, dataset['label_encoded'])

# **3. Pembangunan Model Deep Learning**

Pilihan algoritma pelatihan ini haruslah sesuai dengan tujuan analisis sentimen yang ingin dicapai.

# Skema 2: CNN dengan Word2Vec (80/20)

**Algoritma:** Convolutional Neural Network (CNN)

**Ekstraksi Fitur:** Word2Vec

**Alasan Pemilihan:**
- CNN dikenal dalam pengenalan pola spasial, dan bisa dimanfaatkan untuk menangkap pola lokal pada teks (misalnya n-grams).
- Word2Vec menghasilkan representasi vektor dari kata-kata dengan mempertimbangkan hubungan semantik dan sintaksis.
- Kombinasi CNN + Word2Vec memungkinkan model mempelajari fitur spasial dari representasi kata yang lebih kaya.

In [17]:
# Membagi data menjadi training dan testing
X_train, X_test, y_train, y_test = train_test_split(X_resampled, y_resampled, test_size=0.2, random_state=42)

In [18]:
# Membuat model Word2Vec menggunakan teks bersih
word2vec = Word2Vec(sentences=[text.split() for text in dataset['cleaned_text']], vector_size=100, window=5, min_count=1)

In [19]:
# Menyiapkan embedding matrix
embedding_matrix = np.zeros((len(tokenizer.word_index) + 1, 100))
for word, i in tokenizer.word_index.items():
    if word in word2vec.wv:
        embedding_matrix[i] = word2vec.wv[word]

In [20]:
# Membangun model CNN
model2 = Sequential([
    Embedding(len(tokenizer.word_index) + 1, 100, weights=[embedding_matrix], input_length=100, trainable=False),
    Conv1D(128, 5, activation='relu'),
    GlobalMaxPooling1D(),
    Dense(64, activation='relu'),
    Dropout(0.5),
    Dense(3, activation='softmax')
])

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/embedding.py:90: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


In [21]:
model2.compile(loss='sparse_categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

In [22]:
model2.fit(X_train, y_train, epochs=80, validation_data=(X_test, y_test), batch_size=64)

Epoch 1/80
2392/2392 ━━━━━━━━━━━━━━━━━━━━ 116s 46ms/step - accuracy: 0.7494 - loss: 0.6190 - val_accuracy: 0.8440 - val_loss: 0.3897
Epoch 2/80
2392/2392 ━━━━━━━━━━━━━━━━━━━━ 142s 46ms/step - accuracy: 0.8359 - loss: 0.4196 - val_accuracy: 0.8638 - val_loss: 0.3493
Epoch 3/80
2392/2392 ━━━━━━━━━━━━━━━━━━━━ 157s 52ms/step - accuracy: 0.8556 - loss: 0.3729 - val_accuracy: 0.8774 - val_loss: 0.3172
Epoch 4/80
2392/2392 ━━━━━━━━━━━━━━━━━━━━ 132s 48ms/step - accuracy: 0.8664 - loss: 0.3488 - val_accuracy: 0.8786 - val_loss: 0.3077
Epoch 5/80
2392/2392 ━━━━━━━━━━━━━━━━━━━━ 137s 46ms/step - accuracy: 0.8774 - loss: 0.3228 - val_accuracy: 0.8881 - val_loss: 0.2925
Epoch 6/80
2392/2392 ━━━━━━━━━━━━━━━━━━━━ 154s 51ms/step - accuracy: 0.8827 - loss: 0.3106 - val_accuracy: 0.8799 - val_loss: 0.2979
Epoch 7/80
2392/2392 ━━━━━━━━━━━━━━━━━━━━ 134s 48ms/step - accuracy: 0.8866 - loss: 0.3024 - val_accuracy: 0.8938 - val_loss: 0.2808
Epoch 8/80
2392/2392 ━━━━━━━━━━━━━━━━━━━━ 112s 47ms/step - accuracy: 

In [23]:
# Evaluasi model pada data testing
# Prediksi pada data uji
y_pred1 = np.argmax(model2.predict(X_test), axis=1)

# Laporan Klasifikasi
print("\nClassification Report:")
print(classification_report(y_test, y_pred1, target_names=label_encoder.classes_))

# Akurasi
accuracy = accuracy_score(y_test, y_pred1)
print(f"Akurasi Model: {accuracy:.4f}")

1196/1196 ━━━━━━━━━━━━━━━━━━━━ 12s 10ms/step

Classification Report:
              precision    recall  f1-score   support

     negatif       0.89      0.93      0.91     12770
      netral       0.92      0.96      0.94     12699
     positif       0.95      0.85      0.90     12793

    accuracy                           0.92     38262
   macro avg       0.92      0.92      0.92     38262
weighted avg       0.92      0.92      0.92     38262

Akurasi Model: 0.9168


In [34]:
# Inference dan Testing Skema 2
print("\nInference Skema 2: CNN dengan Word2Vec (80/20)")
# Fungsi Prediksi Teks
def predict_text(text, tokenizer, model2, label_encoder):
    # Tokenisasi dan padding teks
    sequence = tokenizer.texts_to_sequences([text])
    padded_sequence = pad_sequences(sequence, maxlen=100, padding='post')

    # Prediksi sentimen
    prediction = model2.predict(padded_sequence)
    label_index = np.argmax(prediction, axis=1)[0]
    label = label_encoder.inverse_transform([label_index])[0]

    print(f"Teks: {text}")
    print(f"Prediksi Sentimen: {label}")

# Prediksi Teks Baru
print("\nPrediksi Teks Baru:")
predict_text("As an egyptian, I didn't know about the Sphinx.", tokenizer, model2, label_encoder)


Inference Skema 2: CNN dengan Word2Vec (80/20)

Prediksi Teks Baru:
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 54ms/step
Teks: As an egyptian, I didn't know about the Sphinx.
Prediksi Sentimen: netral


**Insight Skema 2 :**

1. Model ini memiliki akurasi yang sangat baik, yaitu 91.68%, menunjukkan kemampuan model dalam mengklasifikasikan sentimen dengan cukup akurat.

2. Kinerja Model pada Setiap Kelas

- Negatif: Precision 0.89, Recall 0.93, F1-Score 0.91
- Netral: Precision 0.92, Recall 0.96, F1-Score 0.94
- Positif: Precision 0.95, Recall 0.85, F1-Score 0.90
- Model memiliki keseimbangan performa yang baik di ketiga kelas, dengan sedikit penurunan pada recall kelas "positif".
3. Pada teks "As an Egyptian, I didn't know about the Sphinx.", model memprediksi sentimen sebagai "netral", yang memang sesuai dengan konteks kalimat.